### ISSUE TITLE
panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.ConfigMap

### ISSUE THREAD

**Issue by monkeymorgan** [jcmoraisjr/haproxy-ingress#932]
**Description of the problem**

During a period where the API server was overloaded (but still functional) and after a leader election in etcd (possible red herring) a large number of instances (150 out of 175) started to panic / crash / fail readiness checks ( /healthz) / restart.

The key log message is : 

panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.ConfigMap

and the key parts of the stacktrace are: 

/go/pkg/mod/k8s.io/client-go@v0.19.13/tools/cache/controller.go
/go/pkg/mod/k8s.io/client-go@v0.19.13/tools/cache/shared_informer.go:779
https://github.com/kubernetes/client-go/blob/v0.19.13/tools/cache/shared_informer.go#L779

It seems that processorListener deleteNotification returns cache.DeletedFinalStateUnknown which is unexpected and causes a panic / lock-up. 

In-flight requests are not completed, and the clients of the in-flight requests timeout.

The readiness probes fail to establish a connection : 

Readiness probe failed: Get "http://10.12.228.150:10253/healthz": dial tcp 10.12.228.150:10253: connect: connection refused

k8s then starts new pods to replace these ones. In our case, this made the load on the API server rocket, as 150 new proxies attempted to list 500+ resources (ingresses / endpoints / secrets / configmaps / pods). I think these namespaces may have 'old' configmaps too which are not cleaned up.

Curiously, only one out of many (similar) deployments collapsed. Others continued to work fine (and emit traces about slow api server syncs) despite the load on the API server rising to the point where it was 95% 5xxs.

**Expected behavior**

Total resilience in the face of all API server issues ! (I know, a lot to ask) 

To reduce the cascade effect, I wonder would it be possible to add the option for label selectors to more tightly control the lister-watchers (observers) ? e.g. in the case of configmap, and in our case we have a 1:1 relationship from proxy to pod deployment so we could ideally only watch pods of interest. This namespace has circa 1000 pods, and 1000s are not unusual.

**Steps to reproduce the problem**

This could be tricky ... I expect it's a race cond. But the only prior errors are:

client-go/leaderelection.go:325
error retrieving resource lock A-NAMESPACE/ingress-controller-leader-haproxy-ingress-redacted-prod: etcdserver: leader changed

1.  Load the api server 
2.  Perhaps force etcdserver leader change ? (I can't tell if this is the cause, but it happened 1 second before the collapse)

**Environment information**

HAProxy Ingress version: `v0.13.7` (with haproxy 2.2.14)

I noticed, that due to a custom build I have k8s go-client 1.19 

Command-line options:

```
    - --configmap=REDACTED-ingress-controller
    - --ingress-class=REDACTED-ingress
    - --sort-backends
    - --watch-namespace=REDACTED
    - --force-namespace-isolation
    - --publish-service=REDACTED
```

Global options:

```
(paste global configmap options here)
```

Ingress objects:

```
nothing 'interesting' in the ingress objects. basic http. 
```

**Comment by jcmoraisjr** [comment]
Hi thanks for the report and a thoroughly description of the problem. `cache.DeletedFinalStateUnknown` struct is already known by the controller but unfortunately a couple of issues contributed together for the downtime:

- only ingress resource has that protection, configmap doesn't have;
- filters aren't thoroughly implemented over all the resource types, so we sometimes process more than we need;
- the controller event loop doesn't recover after a panic.

What I can say for now is that all of these issues are addressed in future versions, but we can do a bit more about it right now, starting with v0.12 and newer, which is check the resource type, since we're aware that network/api failures might lead to ghosts instead of real objects in the internal cache. I'll issue a new couple of tags shortly addressing this.

**Comment by monkeymorgan** [comment]
Thanks for the quick pick-up. We'll certainly adopt as soon as available.

**Comment by jcmoraisjr** [comment]
Fixed in v0.13.8, thanks again for the detailed report! Closing.
